packages/postcss-tape/src/index.ts

```diff
@@ -0,0 +1,344 @@
+
+			// Can't do further checks if "expect" is missing.
+			if (expected === false) {
+				continue;
+			}
+
+			// Assert result with recent PostCSS.
+			//
+			// NOTE:
+			// The version we declare as a peer dependency in plugins.
+			{
+				try {
+					assert.strictEqual(resultString, expected);
+				} catch (err) {
+					hasErrors = true;
+
+					if (process.env.GITHUB_ACTIONS) {
+						console.log(formatGitHubActionAnnotation(
+							formatCSSAssertError(testCaseLabel, testCaseOptions, err, true),
+							'error',
+							{ file: normalizeFilePathForGithubAnnotation(expectFilePath), line: 1, col: 1 },
+						));
+					} else {
+						console.error(formatCSSAssertError(testCaseLabel, testCaseOptions, err));
+					}
+				}
+			}
+
+			// Assert result sourcemaps with recent PostCSS.
+			{
+				try {
+					if (result.map.toJSON().sources.includes('<no source>')) {
+						throw new Error('Sourcemap is broken');
+					}
+				} catch (err) {
+					hasErrors = true;
+
+					if (process.env.GITHUB_ACTIONS) {
+						console.log(formatGitHubActionAnnotation(
+							`${testCaseLabel}\n\nbroken source map: ${JSON.stringify(result.map.toJSON().sources)}`,
+							'error',
+							{ file: testFilePath, line: 1, col: 1 },
+						));
+					} else {
+						console.error(`\n${testCaseLabel}\n\nbroken source map: ${JSON.stringify(result.map.toJSON().sources)}\n\n${dashesSeparator}`);
+					}
+				}
+			}
+
+			// Run "after" when initial postcss run has completely.
+			if (testCaseOptions.after) {
+				await testCaseOptions.after();
+			}
+
+			// Assert that the result can be passed back to PostCSS and still parses.
+			{
+				try {
+					const resultContents = await fsp.readFile(resultFilePath, 'utf8');
+					const secondPassResult = await postcss([noopPlugin()]).process(resultContents, {
+						from: resultFilePath,
+						to: resultFilePath,
+						map: {
+							inline: false,
+							annotation: false,
+						},
+					});
+
+					if (secondPassResult.warnings().length) {
+						throw new Error('Unexpected warnings on second pass');
+					}
+				} catch (_) {
+					hasErrors = true;
+
+					if (process.env.GITHUB_ACTIONS) {
+						console.log(formatGitHubActionAnnotation(
+							`${testCaseLabel}\n\nresult was not parsable with PostCSS.`,
+							'error',
+							{ file: expectFilePath, line: 1, col: 1 },
+						));
+					} else {
+						console.error(`\n${testCaseLabel}\n\nresult was not parsable with PostCSS.\n\n${dashesSeparator}`);
+					}
+				}
+			}
+
+			// Assert result with oldest supported PostCSS.
+			// Check against the actual result to avoid duplicate warnings.
+			//
+			// NOTE:
+			// The oldest version of PostCSS we support at this time.
+			// This does not need to be a different version than the latest version of PostCSS.
+			// There is no system behind setting this.
+			// It is here to allow testing a specific older version, if parts of the community can't update yet.
+			{
```

Comment:
I completely and wholeheartedly love this feature. Thanks for thinking so much ahead of time. ❤️

That said, should we check versions before just in case we already ran this? I see no benefit in running this 3 times with the same version. We'd be "wasting" GH resources and eventually could be capped out.

I reckon this should be a first start: https://postcss.org/api/#processor-version ?

Reply:
I totally forgot we can do that!
Great idea!

Reply:
This has been implemented :)